Accept dune-file as dune file name

nojb opened the pull request:
As discussed in #4386 we decided to allow an alternative name for dune files. As a first step, the present PR allows `dune-file` as an alternative to `dune` (with `dune-file` taking precedence over `dune`). A couple of questions:

- do we version this change?

- what about the documentation? do we publicize `dune-file` as the "official" name from now on?

Fixes #4386

jeremiedimino commented:
> do we version this change?

Yes, `dune-file` should be accepted only in projects using land dune >= 3.0, so that user know in advance that their project won't build with an older Dune.

> what about the documentation? do we publicize dune-file as the "official" name from now on?

Let's talk about this point at the next meeting. Could you add this to the agenda?

nojb commented:
> > do we version this change?
> 
> Yes, `dune-file` should be accepted only in projects using land dune >= 3.0, so that user know in advance that their project won't build with an older Dune.

OK, for now I changed the PR so that under `>= 3.0` only `dune-file` is accepted and implemented a basic support for `dune upgrade` to do the renaming automatically, but we can discuss this more in-depth in the next meeting.

> > what about the documentation? do we publicize dune-file as the "official" name from now on?
> 
> Let's talk about this point at the next meeting. Could you add this to the agenda?

Done!

jeremiedimino commented on src/dune_engine/source_tree.ml:
```diff
@@ -491,8 +493,24 @@ end = struct
+        if String.Set.mem files Dune_file.dune_fname then
+          Some Dune_file.dune_fname
+        else
+          None
```
Might be nice to print a warning if the `dune-file` exists, saying that it is only supported sincee 3.0

nojb commented:
We discussed this today at the dev meeting and we agreed that we will accept an alternative file name in 3.0 and later as an opt-in feature (to be enabled in `dune-project`). `dune` will continue to be accepted as a file name in any case.

nojb commented:
I pushed a revised version, with tests. Review welcome. Thanks!

jeremiedimino reviewed (approved):
Looks good!

nojb commented:
Thanks for the review. I pushed some doc and am planning to merge once CI is green.
